gramps_webapi/api/resources/file.py

```diff
@@ -0,0 +1,66 @@
+"""Endpoint for up- and downloading media files."""
+
+import json
+from http import HTTPStatus
+
+from flask import Response, abort, current_app, request
+from gramps.gen.db import DbTxn
+from gramps.gen.errors import HandleError
+
+from ...auth.const import PERM_EDIT_OBJ
+from ..auth import require_permissions
+from ..file import LocalFileHandler, process_file, upload_file
+from ..util import get_db_handle
+from . import ProtectedResource
+from .util import transaction_to_json, update_object
+
+
+class MediaFileResource(ProtectedResource):
+    """Resource for media files."""
+
+    def get(self, handle) -> Response:
+        """Download a file."""
+        db_handle = get_db_handle()
+        try:
+            obj = db_handle.get_media_from_handle(handle)
+        except HandleError:
+            abort(HTTPStatus.NOT_FOUND)
+        base_dir = current_app.config.get("MEDIA_BASE_DIR")
+        handler = LocalFileHandler(handle, base_dir)
+        return handler.send_file(etag=obj.checksum)
+
+    def put(self, handle) -> Response:
+        """Upload a file and update the media object."""
+        require_permissions([PERM_EDIT_OBJ])
+        db_handle = get_db_handle()
+        try:
+            obj = db_handle.get_media_from_handle(handle)
+        except HandleError:
+            abort(HTTPStatus.NOT_FOUND)
+        checksum_old = obj.checksum
+        for etag in request.if_match:
+            if etag != checksum_old:
+                abort(412)
+        mime = request.content_type
+        if not mime:
+            abort(HTTPStatus.NOT_ACCEPTABLE)
+        checksum, f = process_file(request.stream)
+        if checksum == obj.checksum:
+            # don't allow PUTting if the file didn't change
+            abort(HTTPStatus.CONFLICT)
+        base_dir = current_app.config.get("MEDIA_BASE_DIR")
+        path = upload_file(base_dir, f, checksum, mime)
+        obj.set_checksum(checksum)
+        obj.set_path(path)
+        obj.set_mime_type(mime)
+        db_handle_writable = get_db_handle(readonly=False)
+        with DbTxn("Update media object", db_handle_writable) as trans:
+            update_object(db_handle_writable, obj, trans)
+            try:
+                pass  # update_object(db_handle_writable, obj, trans)
+            except ValueError:
+                abort(400)
+            trans_dict = transaction_to_json(trans)
+        return Response(
+            response=json.dumps(trans_dict), status=200, mimetype="application/json"
+        )
```

Comment:
That's true, currently there is no way to delete files and there is no automatic cleanup. There is a possibility to delete *media objects* but that does not remove the corresponding file.

This is similar to how it works in Gramps: deleting a Gramps media object does not delete the corresponding file. Gramps actually never deletes files.

This is also why an automatic cleanup cannot really work: just because there is a file in the media base dir that does not have a corresponding media object doesn't mean we can delete it (I have plenty of files in my media directory that do not have associated objects).

There are two problems with this:

1. The media directory might keep filling up with outdated/unused files
2. An inadvertent upload of a file (e.g. a user uploads an image and then realizes the image is copyrighted or otherwise private, and tries to delete it again) cannot be undone

I think 1. has to be solved manually by the administrator, I don't see any other way. I don't see a way to solve 2. at all, so I guess users have to be careful.

Reply:
For 1 you can scan the media dir and cross reference against the media objects, so you could identify orphans to provide a report or allow the admin to schedule an automated cleanup of say anything orphaned and more than 30 days old.  

I see it as different from desktop Gramps where you have control over the file name and location and you are the one adding them and deleting them.

For 2 you would need to add Contributor to the Gramps data model and track physical file contributions separate of the media object, or add such tracking to the user database on the API side.

Not something needed now but to keep in mind I guess. As I understand it I think 2 must be addressed for GDPR.

Reply:
> As I understand it I think 2 must be addressed for GDPR.

I think the possibility to delete content (directly) and the right for deletion of personal data are separate - even if the user cannot delete a file they can ask the tree owner to delete anything that contains their personal data.

> For 1 you can scan the media dir and cross reference against the media objects, so you could identify orphans to provide a report or allow the admin to schedule an automated cleanup of say anything orphaned and more than 30 days old.

That sounds like a great idea for a Gramps (desktop) addon. There are already several cleanup tools for other object types. The admin could use that regularly with their synchronised database & media directory.

Btw, I also want to re-implement the S3 file handler I had in the "old" app at some point - in that case, only files with a corresponding media object will be stored on S3 with the checksum as handle, and I don't see a problem with deleting them.